Added blog post for announcement - draft (#76)
premAI-io/dev-portal: blog/2023-11-06-towards-censorship-resistant-architecture/index.md

Reviewer: @filopedraz should we add references to [State of Open Source AI Book - 2023 Edition](https://book.premai.io/state-of-open-source-ai)?

Author: Good point. Yes. We should, but where would you put it?

Reviewer: We can either use it to introduce Prem and its contribution to open source trough  the book or reference chapters when we talk about fine tuning, open-source and so on. Or even both if it's not too much

Author: Yep, good idea.

---
slug: towards-censorship-resistant-architecture
title: Towards a Censorship Resistant Architecture
authors: [filippopedrazzinfp]
tags: [llm, self-hosted, prem, open-source, petals, p2p, decentralized-ai]
description: "Prem - Towards a Censorship Resistant Architecture"
image: "./banner.png"
---

![Prem Banner](./banner.png)

> Horizontal Scaling is all you need.

> Prem Labs is introducing **Prem Network**, a decentralized AI Platform enabling consumer devices to run Large Language Models without compromising on quality, aggregating the computational power of devices around the globe. Prem Network is a breakthrough in LLM inference deployment as it enables consumer devices to power GPU-intensive models on the edge. Similar to what BitTorrent did for P2P file sharing, Prem Swarm provides AI inference and training for consumer devices.

Author: `Prem Swarm` is the old name. We should replace it everywhere with `Prem Network`.

> Prem Labs is introducing **Prem Network**, a decentralized AI Platform enabling consumer devices to run Large Language Models without compromising on quality, aggregating the computational power of devices around the globe. Prem Network is a breakthrough in LLM inference deployment as it enables consumer devices to power GPU-intensive models on the edge. Similar to what BitTorrent did for P2P file sharing, Prem Swarm provides AI inference and training for consumer devices.

Reviewer: @filopedraz is it only for LLM or could it be more general (text-to-image, text-to-video, ...)?

Author: For now only LLM.

They are the first step towards creating an alternative to OpenAI centralization.
However, as more powerful open-source models and weights become available to the public, there is still a huge gap between the requirements necessary to run these models and the current computing capabilities of consumer hardware. Today, consumer hardware is still far from leveraging these models' capabilities (e.g., Llama 65B requires around 90GiB of RAM). For this reason, projects like llama.cpp, Alpaca, Vicuna, and GPT4All emerged to run smaller, distilled, quantized versions of these models using consumer devices (CPU and GPU).

On the one hand, this proved that using different methodologies like quantization and fine-tuning smaller models on very clean datasets could reach incredible relative results. On the other hand, these models are still far from ChatGPT and GPT4 quality and semantic understanding.

The low quality is pushing further research to improve these "small" models and increase quality with new methods and techniques to tackle the main issues:

- Context dimension
- Generation quality and hallucination

> Some very good papers have come out lately to face these issues (e.g., X, Y), but no major breakthrough has yet emerged. It would seem that perhaps the Transformer model architecture optimization itself is not good enough for the next generation of LLMs with consumer hardware constraints. Only research will be able to answer these open questions.

Author: No list. Two options: (1) we look them up, (2) we remove this strong sentence. Wdyt?

At Prem, we are focusing on engineering a new paradigm of decentralized and privacy-preserving AI - AI that is always yours. As we push the boundary of what is possible in the deployment of Open Source models at the edge, we are excited to offer an alternative solution to the problem of efficient and optimized deployment of AI models ad the edge.

> As Prem Labs we cannot do bets on research, but only engineering. We can go low level and tackle all the engineering issues step by step solving one problem after the other, but we cannot build a startup based on a research bet.

On the other hand, consumer hardware will be more and more powerful. They are probably able to run bigger and bigger models.

> > Similar as above, at Prem Labs we cannot wait that hardware gets better and better of few magnitudes orders before delivering quality to our end users. It's all about balance and bets and even if things are going crazily fast, still ChatGPT needs 175B Parameters to be the model that it is now. Assuming averagely 1GiB per RAM for each 1B Parameters, we are talking about 175GiB of Memory in order to run the state of the art known model. Obviously, even the best consumer device is far from years to be able to have this amount of memory and we really cannot this to happen.

Reviewer 2: ```suggestion
> Similarly, at Prem Labs we cannot wait till hardware gets better by few orders of magnitude before delivering quality to our end users. For us it's all about balance and bets and even if things are going crazily fast, still ChatGPT needs 175B Parameters to be the model that it is now. Assuming on average 1GiB of VRAM for each 1B Parameters, we are talking about 175GiB of GPU Memory in order to run the state of the art known model. Obviously, even the best consumer device is far from years to have this amount of memory.
```

## Twitter vs. Reality

> The main problem right now is that we don't have a clear benchmark to use in order to evaluate these models. Both Open LLM Leaderboard and the Alpaca Leaderboard are still far from being able to properly evaluate this new family of LLMs.

Reviewer 2: ```suggestion
The main problem right now is that we don't have a clear benchmark to use in order to evaluate these models. Both [Open LLM Leaderboard](https://huggingface.co/spaces/HuggingFaceH4/open_llm_leaderboard) and the [Alpaca Leaderboard](https://tatsu-lab.github.io/alpaca_eval/) are still [far from being able to properly evaluate this new family of LLMs](https://dev.premai.io/blog/evaluating-open-source-llms).
```

> A lot of times happened in the last month that a new model reached the top of the Open Source LLM Leaderboard and then after few simple empirical tests you get disappointed by the obvious bad results.

Reviewer 2: ```suggestion
A lot of times we have seen that a new model reaches the top of the Open Source LLM Leaderboard and then after few simple empirical tests by someone, producing disappointing results.
```

As an industry, we are doing a lot of steps and improvements on a daily basis, but we are not there yet and sometimes the hype confuses what is real and what is not.

## Smaller Models for Smaller Tasks

> Another movement that could be considered more like an old way of thinking is to use smaller, specialized models to handle specific tasks. A phenomenon that has not been seen yet in Production so far is to leverage these LLMs for already famous/mainstream NLP tasks like summarization, sentiment analysis, etc. Fine-tuning an LLM and exploiting the semantic understanding of these models to solve atomic NLP tasks, could be another movement that we will see more and more in the future. On the other hand, it's still soon to see these models in Production. A state-of-the-art BERT used for sentiment analysis for example generates responses under 500ms and can be used in a real-time use case. A Llama-7b model 4-bit quantized and fine-tuned on sentiment analysis tasks is still far from this latency requirement even if performances would be probably better in terms of quality.

Reviewer 2: ```suggestion
Another movement that could be considered is to use smaller and specialized models to handle specific tasks. A phenomenon that has not been seen yet in Production so far is to leverage these LLMs for most used NLP tasks like summarization, sentiment analysis, classification, etc. Fine-tuning an LLM and exploiting the semantic understanding of these models to solve atomic NLP tasks, could be another movement that we will see more and more in the future. On the other hand, it's still early to see these models in Production. A state-of-the-art BERT used for sentiment analysis for example generates responses under 500ms and can be used for real-time use cases. A Llama-7B 4-bit quantized model fine-tuned on sentiment analysis tasks is still far from this latency requirement, even if it performs much better in terms of quality.
```

> ## A new Paradigm can Quickly Take Over

Reviewer 2: ```suggestion
## A new Possible Paradigm
```

Author: Not sure if it's an improvement.
